ICE: `Result::unwrap()` on an `Err` value: Utf8Error

Compiler crashes while compiling https://github.com/CrystalGamma/rusttype/tree/bugrep0 using Cargo.

Cargo version: cargo 0.6.0-nightly (18e496a 2015-10-17)
rustc version: 1.3.0

     thread 'rustc' panicked at 'called `Result::unwrap()` on an `Err` value: Utf8Error { valid_up_to: 502 }', src/libcore/result.rs:732

    stack backtrace:
       1:     0x7fc6f8f4766e - sys::backtrace::write::ha67a6a2049d22fadIvs
       2:     0x7fc6f8f4fd55 - panicking::on_panic::hfa4bf0215983f759lzx
       3:     0x7fc6f8f1094e - rt::unwind::begin_unwind_inner::ha1b5da93bdf7e8ffC0w
       4:     0x7fc6f8f116b6 - rt::unwind::begin_unwind_fmt::hac9e09b20e42a51cIZw
       5:     0x7fc6f8f4f8f6 - rust_begin_unwind
       6:     0x7fc6f8f96b64 - panicking::panic_fmt::hd75b6cee85f3afe4SgC
       7:     0x7fc6f7cf58cd - back::link::link_natively::hed2bd12732711d7db0b
       8:     0x7fc6f7cecded - back::link::link_binary::h7266bf96d64e0248uvb
       9:     0x7fc6f94ba8cd - driver::phase_6_link_output::h0d7d5071f34ef57fiRa
      10:     0x7fc6f948ce44 - driver::compile_input::h52c7dafd49963360Tba
      11:     0x7fc6f9572a63 - run_compiler::h473c62e00f865fa9A7b
      12:     0x7fc6f95704de - boxed::F.FnBox<A>::call_box::h10713159552398332324
      13:     0x7fc6f956fe09 - rt::unwind::try::try_fn::h16887904248528792425
      14:     0x7fc6f8f4f84d - __rust_try
      15:     0x7fc6f8f3a8f7 - rt::unwind::try::inner_try::hafffff77ddacfa5fvWw
      16:     0x7fc6f9570028 - boxed::F.FnBox<A>::call_box::h5417278073590676257
      17:     0x7fc6f8f4e991 - sys::thread::Thread::new::thread_start::h29dca2a2cf2294b535v
      18:     0x7fc6f37c64a3 - start_thread
      19:     0x7fc6f8ba813c - clone
      20:                0x0 - <unknown>

``note: zzz\xff/libfoo.so: undefined reference to `foo\'\ncollect2: error: ld returned 1 exit status\n``

Not great, but not an ICE and probably good enough until we have a dedicated function to do this sort of thing since this should be a very rare error.

--- a/src/librustc_trans/back/link.rs
+++ b/src/librustc_trans/back/link.rs
@@ -32,6 +32,8 @@ use util::sha2::{Digest, Sha256};
 use util::fs::fix_windows_verbatim_for_gcc;
 use rustc_back::tempdir::TempDir;
 
+use std::ascii;
+use std::char;
 use std::env;
 use std::ffi::OsString;
 use std::fs::{self, PathExt};
@@ -883,6 +885,16 @@ fn link_natively(sess: &Session, dylib: bool,
     let prog = time(sess.time_passes(), "running linker", || cmd.output());
     match prog {
         Ok(prog) => {
+            fn escape_string(s: &[u8]) -> String {
+                str::from_utf8(s).map(|s| s.to_owned())
+                    .unwrap_or_else(|_| {
+                        let mut x = "Non-UTF-8 output: ".to_string();
+                        x.extend(s.iter()
+                                 .flat_map(|&b| ascii::escape_default(b))
+                                 .map(|b| char::from_u32(b as u32).unwrap()));
+                        x
+                    })
+            }
             if !prog.status.success() {
                 sess.err(&format!("linking with `{}` failed: {}",
                                  pname,
@@ -890,11 +902,11 @@ fn link_natively(sess: &Session, dylib: bool,
                 sess.note(&format!("{:?}", &cmd));
                 let mut output = prog.stderr.clone();
                 output.push_all(&prog.stdout);
-                sess.note(str::from_utf8(&output[..]).unwrap());
+                sess.note(&*escape_string(&output[..]));
                 sess.abort_if_errors();
             }
-            info!("linker stderr:\n{}", String::from_utf8(prog.stderr).unwrap());
-            info!("linker stdout:\n{}", String::from_utf8(prog.stdout).unwrap());
+            info!("linker stderr:\n{}", escape_string(&prog.stderr[..]));
+            info!("linker stdout:\n{}", escape_string(&prog.stdout[..]));
         },
         Err(e) => {
             sess.fatal(&format!("could not exec the linker `{}`: {}", pname, e));

